### PR TITLE
Switch to vibe-d thread-local connection pool, remove shared

### PR DESCRIPTION
**Pull request by Boris-Barboris**
This PR changes PostgresClient in order to use vibe-d built-in semaphore connection pool implementation, which is thread-local and therefore can no longer be shared. 

Interface of LockedConnection was left intact, but absence of shared may still break client code.

Also, do you mind .gitignore here, or you prefer them out of the repo?

### CODE REVIEW

File: source/vibe/db/postgresql/package.d [hunk 1]
```diff
@@ -39,25 +41,37 @@ shared class PostgresClient
...
-
-        LockedConnection!__Conn conn;
-
-        while(!pool.tryLockConnection(conn))
-        {
-            yield(); // TODO: Need to replace yield() by events handling in the conn pool for compatibility with threads and fibers both
-            continue;
-        }
+        logDebugV("get connection from the pool");
+        return new LockedConnection!__Conn(pool.lockConnection());
```
Comment (denizzzka) [inline source/vibe/db/postgresql/package.d]: What will happen If here is no free connections in the pool?

Comment (Boris-Barboris) [reply source/vibe/db/postgresql/package.d]: fiber will block on the userspace semaphore.

Comment (Boris-Barboris) [reply source/vibe/db/postgresql/package.d]: https://github.com/vibe-d/vibe-core/blob/master/source/vibe/core/connectionpool.d#L76

vibe-d:core should be allright, vibe-core had a bug in semaphore until https://github.com/vibe-d/vibe-core/pull/31

Comment (denizzzka) [reply source/vibe/db/postgresql/package.d]: Memo: do not forgot to increase vibe version

**Comment by denizzzka** [comment]
Old **shared** connections pool saved to:
https://gist.github.com/denizzzka/b1e0c199c5df46840610e3c45aed975a
(maybe someone will need it in the future)

**Comment by denizzzka** [comment]
> Also, do you mind .gitignore here

Yes

**Comment by Boris-Barboris** [comment]
kebab removed

**Comment by Boris-Barboris** [comment]
@denizzzka 
can you look at ci log? I simply removed .gitignore, in the last commit, and it broke.

**Comment by denizzzka** [comment]
> kebab removed

Sorry, I am not native English language speaker. My "yes" was about: ".gitignore should be included into this package"

**Comment by denizzzka** [comment]
Why `example` subpackage builds ok? Hmm...

**Comment by Boris-Barboris** [comment]
fixed

**Comment by denizzzka** [comment]
Sorry, but forgotten one small change: need to set up new vibe.d version in dub.sdl

**Comment by denizzzka** [comment]
(new vibe-core with your fixes released today)

**Comment by Boris-Barboris** [comment]
This way allright? Or you think it's better to add configuration vibe-core or something. End choice is done by dub, if vibe-d:core is build in "vibe-core" configuragion, it is used. I'm not sure dub will respect optional dependency's version restriction.

**Comment by denizzzka** [comment]
> This way allright?

no :-(

Ok revert 8c32ced and I will merge this PR.

**Comment by Boris-Barboris** [comment]
yeah, some strange stuff:
https://github.com/dlang/dub/issues/1089

**Comment by denizzzka** [comment]
> yeah, some strange stuff:
> dlang/dub#1089

So, we can't move to new vibe-core because old vibe-core used by... dpq2?

**Comment by Boris-Barboris** [comment]
I don't know. Generally, vibe-d packages have no restrictions on each other versions, so even this:
https://github.com/denizzzka/dpq2/blob/master/dub.json#L13
line should not be a problem. I just couldn't utilize optional dependency correctly, since it was actually used and conflicted with vibe-d:core, wich must be set to specific configuration when vibe-core is present. More of a dub problem than ours.

**Comment by denizzzka** [comment]
This string was added only few weeks ago:
https://github.com/denizzzka/dpq2/commit/6a1ff5628d68488be334e96764120c2b7b1b7794

**Comment by denizzzka** [comment]
> This string was added only few weeks ago:
denizzzka/dpq2@6a1ff56

Tried to remove - isn't compiles with same error. vibe.d inside of it uses old vibe core.

vibe:core also depends from vibe-core version="~>1.0"

Hmm....

**Comment by Boris-Barboris** [comment]
> vibe:core also depends from vibe-core version="~>1.0"

I guess we'll need to PR vibe-d.

**Comment by denizzzka** [comment]
https://68.media.tumblr.com/0642328fb18baf588fa88cbec6d26851/tumblr_ntygfmWVdE1udwanoo1_1280.jpg
image

**Comment by Boris-Barboris** [comment]
https://github.com/thepracticaldev/orly-full-res/blob/master/resolvingbrokendependencies-big.png

**Comment by denizzzka** [comment]
Fixed by adding dub.selection.json

**Comment by denizzzka** [comment]
@Boris-Barboris: communicated with you and you were interested in supporting of the native connection to Postgres (without libpq). Now, if it is still needed it can implemented here: https://github.com/denizzzka/dpq2/blob/add_native_connection/src/dpq2/native_conn/connection.d

**Comment by Boris-Barboris** [comment]
@denizzzka yeah, the thought is still dwelling. On the dpeq side, i still need to finish a couple of issues: https://github.com/Boris-Barboris/dpeq/issues for the sake of completeness and better type support.
After that, this is indeed the right place to provide middle-level code.

A couple of ideas that come from my personal experience and i'd like to reflect them here, when the time comes:
* These are main connection usage patterns I have encountered:
    1. Simple, relatively-fast readonly queries. No repeated roundtrip, no explicit transactions. Such requests are easily pipelined and multiplexed, and have a very good throughtput even on just one connection (tens of megabytes per second). They also benefit greatly from persistent pre-created prepared statements. Such queries, however, may require something like Future.
    2. Very heavy, long-running readonly statistical queries wich can run for hours and should not be subject to timeouts. They also cannot share connections.
    3. Standard exclusive connection all the pools provide currently, where writes and explicit\implicit transactions are used. Correct error-handling is the most important issue here.
    4. Notify listener, wich is constantly hanging in "read" on the socket and is not subject to read timeouts. This well may be abstracted away by notify-receiving callback on the pool itself, without exposing any objects to the client code.
* We use typical active\passitve psql deployment. Pattern 1 can be routed to any backend. Pattern 2 should be routed to replicating backends if possible. Pattern 3 and 4 must be ran on the active nodes. I think connection pool should be aware of active\passive postgres cluster and provide different flavors of connections (at least for patterns 1-3).

Your thoughts?

**Comment by denizzzka** [comment]
0. I am propose to implement native connections inside of `dpq2`, not `vibe.d.db.postgresql`. This is two different things, `vibe.d.db.postgresql` depends from `dpq2` (I am just found your previous messages in this discussion)

> Simple, relatively-fast readonly queries. No repeated roundtrip, no explicit transactions.

Why? Transactions are trivial to implement. For example: https://github.com/DSoftOut/pgator/blob/master/source/sql_transaction.d

What is difference between i and iii?

> They also benefit greatly from persistent pre-created prepared statements. Such queries, however, may require something like Future.

Sure. This is just yet another way to implement async queries.

> Very heavy, long-running readonly statistical queries wich can run for hours and should not be subject to timeouts.

Can it be implemented as optional Timeout == 0?

> They also cannot share connections

No one can share connections in `vibe.d.db.postgresql`

> Notify listener, wich is constantly hanging in "read" on the socket and is not subject to read timeouts. This well may be abstracted away by notify-receiving callback on the pool itself, without exposing any objects to the client code.

Yes.

> We use typical active\passitve psql deployment. Pattern 1 can be routed to any backend. Pattern 2 should be routed to replicating backends if possible. Pattern 3 and 4 must be ran on the active node. I think connection pool should be aware of active\passive postgres cluster and provide different flavors of connections (at least for patterns 1-3).

It is too difficult to define this "roles" for connections and it can be implemented as some number of pools: 1 can be implemented in high level factory of connections which can use connections from another pool.

**Comment by Boris-Barboris** [comment]
> What is difference between i and iii?
> Sure. This is just yet another way to implement async queries.
> No one can share connections in vibe.d.db.postgresql

These three lines are interlinked. Libpq's choice, reflected in the quote "PQsendQuery cannot be called again (on the same connection) until PQgetResult has returned a null pointer, indicating that the command is done." from here: https://www.postgresql.org/docs/current/static/libpq-async.html, and those batching patches we discussed that apparently have never arrived to the libpq's upstream.

Async queries are just queries on non-blocking sockets. I'm talking about command pipelining.
The difference between i and iii is the fact, that user can use small number of connections (one, two tops) to pump all simple one-liner commands through them. It scales much better then naiive approach of just increasing connection pool size, and I generally consider it a beautiful solution.

We are not forced to do it libpq's way. Also, dpeq written in fully synchronous matter, but the socket itself is templated, delegating scheduling to vibe-core.

> Why? Transactions are trivial to implement.

Pipelining is impossible with explicit transactions, hence the segregation into two patterns. Also, I think that PGator example is a bit crude, for example there is no way to specify transaction isolation level. And it's not like it's hard to just send "BEGIN;" command. IMHO transactions, safepoints etc should be left for the user, but it does no harm the other way.

> No one can share connections in vibe.d.db.postgresql
> It is too difficult to define this "roles" for connections and it can be implemented as some number of pools: 1 can be implemented in high level factory of connections which can use connections from another pool.

I do not mind that. Pattern 1 however does not imply explicit connection locking by the client code at all. You just ask the connection pool to execute one-liner and either get a Future wich your task can block on, or block immediately (that is for us to decide, or we can implement both ways). This way we come back to only one type of connection - explicitly locked exclusive one.

**Comment by denizzzka** [comment]
> We are not forced to do it libpq's way. Also, dpeq written in fully synchronous matter, but the socket itself is templated, delegating scheduling to vibe-core, wich may become a problem for dpq2 integration approach.

Isn't problem because 1) it will be new "Native socket" and `dpq2` provides only `Result` navigation and `Value` decoding for it.  2) vibe-core becomes new D's standard events library (I hope)

> Pipelining is impossible with explicit transactions, hence the segregation into two patterns. Also, I think that PGator example is a bit crude, for example there is no way to specify transaction isolation level.

Pgator provides "default" and "read only" transactions levels. It can be implemented another type of transactions by same way. I provide link only as example of how transactions can be implemented. struct guarantees that transaction will be closed and next connection user isn't stucks into non-closed transaction.

> I do not mind that. Pattern 1 however does not imply explicit connection locking by the client code at all. You just ask the connection pool to execute one-liner and either get a Future wich your task can block on, or block immediately (that is for us to decide, or we can implement both ways). This way we come back to only one type of connection - explicitly locked exclusive one.

I intuitively do not like the idea of combining the connection pool and the library to interact with the database.

If there is no pool then there must be a locking until the future isn't completes.
If here is pool implemented it should.. What? What if user do not want to use transactions for each request for some reason? (performance?) But usage of pool without transaction can cause unpleasant side effects.

**Comment by Boris-Barboris** [comment]
> I intuitively do not like the idea of combining the connection pool and the library to interact with the database. If there is no pool...

There is pool. Pool just has additional methods for optimized one-line requests.

> ...then there must be a locking until the future isn't completes.

I did vibe-d semaphore-based request queue. Locking is simply hidden from the client code, but it is there. It's no longer connection-granular for small requests, that's the only difference.

> If here is pool implemented it should.. What? What if user do not want to use transactions for each request for some reason? (performance?) But usage of pool without transaction can cause unpleasant side effects.

Pool may ba a bad word at this point. Connector, maybe? You can ask connector to:
* lock the connection from it's internal pool for you to use exclusively.
* issue transaction-less requests, without managing connection aquisition explicitly (it will be done by the connector internally).

Internally, connector just maintains old-school pools and hides some logic.

**Comment by denizzzka** [comment]
> Pool just has additional methods for optimized one-line requests.

This and pipeline execution can be implemented without additional "pool"

**Comment by Boris-Barboris** [comment]
> This and pipeline execution can be implemented without additional "pool"?
yes

**Comment by denizzzka** [comment]
So maybe it's better to do so? And pool (if anyone need it) can be added at higher level.

**Comment by Boris-Barboris** [comment]
> So maybe it's better to do so? And pool (if anyone need it) can be added at higher level.

vibe.d.db.postgresql seems appropriate, no? Oh, you were objecting because you thought i wanted to do this pooling\pipelining stuff in dpq2?

**Comment by denizzzka** [comment]
> vibe.d.db.postgresql seems appropriate, no? 

Yes, appropriate

> Oh, you were objecting because you thought i wanted to do this pooling\pipelining stuff in dpq2?

Yes.

But I'm not sure about pipelining, maybe it can be implemented in dpq2 for native connections?

**Comment by Boris-Barboris** [comment]
it would require synchronization primitives from vibe-core, so I'm not sure. If possible, I think dpq2 should not be hard-bound to vibe-d.

**Comment by denizzzka** [comment]
Можно сначала привязать а потом как заработает сделать привязку опциональной. Но, похоже,vibe-core таки станет стандартной библиотекой асинхронных эвентов (типа std.event или std.eventloop). Недавно в телеграммном чатике смотрели - вроде бы vibe-core совместима с фобосовыми примитивами.

**Comment by Boris-Barboris** [comment]
@denizzzka Что думаешь на тему если я потихоньку вынесту dpq2.conv в отдельную либу в дальнейшем, а то мне так лень типы пилить. Что-то типо общей либы по типу pgtypes или даже sqltypes было бы круто, я б тогда просто к ней прицепился для всяких нумериков\дат\географии.

На тему въехать dpeq в dpq2 всё-таки боюсь я пас, мне не очень нравится классическая концепция предоставлять юзеру коннекшн. Впилил тут в своё говно на работе вот такую штуку:

https://github.com/Boris-Barboris/pgpepe/blob/master/tests/source/main.d
https://github.com/Boris-Barboris/pgpepe/blob/master/source/pgpepe/connector.d
ну и в целом пощёлкай по https://github.com/Boris-Barboris/pgpepe/tree/master/source/pgpepe плз, расскажи что думаешь.

Я считаю именно такая штука это годный мидл-левел для писателей сервисов, с одним коннекшеном это тупо. И вот уже поверх этого обмазываться орм-кой для особых любителей.

**Comment by denizzzka** [comment]
> Что думаешь на тему если я потихоньку вынесту dpq2.conv в отдельную либу в дальнейшем, а то мне так лень типы пилить.

Без проблем! Это может быть неплохая идея, хотя граница может пройти не точно по `dpq2.conv`- по памяти не помню, надо смотреть внимательнее. (Пишите если что, может можно уточнить границы просто в рамкках пакета dpq2)

Ну и в целом, мой "хитрый план" как раз в том, чтобы кто-то запилил нативные соединения в рамках dpq2 (препятствий для этого, вроде, нет), и, собственно, конвертация бы осталась в рамках `dpq2` потому что мы тут байтоёбствуем во всю.

**Comment by denizzzka** [comment]
> ну и в целом пощёлкай по https://github.com/Boris-Barboris/pgpepe/tree/master/source/pgpepe плз, расскажи что думаешь.

подожди ~сутки-двое, будет ответ

**Comment by Boris-Barboris** [comment]
ну пасматрииии.

https://github.com/Boris-Barboris/pgpepe

вон какие картинки красивые

**Comment by denizzzka** [comment]
Сорян, посмотрел.

А не знаю: концепция быстрых запросов понятная (обсуждали ранее), нативное соединение тоже, но оно же `dpq2` не юзает. А у нас там вылизанная система бинарных типов.

А пул это более высокоуровневая концепция (обсуждали уже когда-то тоже). Для её реализации у обычных (любых - хоть нативных хоть библиотечных) просто должны торчать наружу нужные "ручки"-методы.

**Comment by Boris-Barboris** [comment]
Ну решай. Просто вхуячить коннекшен из dpeq в dpq2 как файлик несложно, сложно подружиться с структурами\классами типа Result, InternalQueryParams, какие-то режимы типа setSingleRowMode - всё очень продиктовано тем как это делает libpq. Можно конечно добавить dpeq в депенденси а в dpq2 нагородить адаптеров. И с сокетом чё делать я так и не понял. Ну или забить и разъехаться по типам, я в своей каше, вы в своей, но я стильну типы.

> А пул это более высокоуровневая концепция (обсуждали уже когда-то тоже)

Ну у меня как сервисописателя вся жизнь только отсюда и начинается, и всё что ниже уровнем интересно только потому что приходится знать. Ну это ладно, дело юзкейса.

> Для её реализации у обычных (любых - хоть нативных хоть библиотечных) просто должны торчать наружу нужные "ручки"-методы.

Это жиденький пул, который не знает нихрена про то что он пулит. Можно лучше, гуще.

> в телеграммном чатике

Чатик?

**Comment by denizzzka** [comment]
> сложно подружиться с структурами\классами типа Result,

Общий интерфейс IResult должен подойти.

> InternalQueryParams,

Нинужен тебе вообще

> какие-то режимы типа setSingleRowMode

Ненужен, потом как-то по-другому запилишь если подобное понадобится.

> - всё очень продиктовано тем как это делает libpq. Можно конечно добавить dpeq в депенденси а в dpq2 нагородить адаптеров.

Не можно а нужно. Оно там не так уж и сложно, веточку даже уже сделал для этого.

> И с сокетом чё делать я так и не понял.

Не имеет значения, о сокетах `dpq2` как бы ничего не знает.
Пул сейчас работает с ними, но если будут у твоего нативного коннекта торчать другие способы придания асинхронности наружу поддержать это в пуле будет тривиально.

> Ну или забить и разъехаться по типам, я в своей каше, вы в своей, но я стильну типы.

Устанешь обновлять.

> в телеграммном чатике

https://t.me/dlangru

**Pull request merged**